daemon/containerd/image_snapshot_unix.go

```diff
@@ -0,0 +1,76 @@
+//go:build !windows
+
+package containerd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/docker/docker/pkg/idtools"
+)
+
+func (i *ImageService) remapSnapshot(ctx context.Context, snapshotter snapshots.Snapshotter, id string, parentSnapshot string, lease leases.Lease) error {
+	ls := i.client.LeasesService()
+	rootPair := i.idMapping.RootPair()
+	usernsID := fmt.Sprintf("%s-%d-%d", parentSnapshot, rootPair.UID, rootPair.GID)
+	remappedID := usernsID + remapSuffix
+
+	// If the remapped snapshot already exist we only need to prepare the new snapshot
+	if _, err := snapshotter.Stat(ctx, usernsID); err == nil {
+		_, err = snapshotter.Prepare(ctx, id, usernsID)
+		return err
+	}
+
+	if err := ls.AddResource(ctx, lease, leases.Resource{
+		ID:   remappedID,
+		Type: "snapshots/" + i.StorageDriver(),
+	}); err != nil {
+		return err
+	}
+	if err := ls.AddResource(ctx, lease, leases.Resource{
+		ID:   usernsID,
+		Type: "snapshots/" + i.StorageDriver(),
+	}); err != nil {
+		return err
+	}
+
+	mounts, err := snapshotter.Prepare(ctx, remappedID, parentSnapshot)
+	if err != nil {
+		return err
+	}
+
+	if err := remapRootFS(ctx, mounts, rootPair); err != nil {
+		snapshotter.Remove(ctx, usernsID) //nolint: errcheck
+		return err
+	}
+
+	if err := snapshotter.Commit(ctx, usernsID, remappedID); err != nil {
+		return err
+	}
+
+	_, err = snapshotter.Prepare(ctx, id, usernsID)
+	return err
+}
+
+func remapRootFS(ctx context.Context, mounts []mount.Mount, idpair idtools.Identity) error {
+	return mount.WithTempMount(ctx, mounts, func(root string) error {
+		return filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			stat := info.Sys().(*syscall.Stat_t)
+			if stat == nil {
+				return fmt.Errorf("cannot get underlying data for %s", path)
+			}
+
+			return os.Lchown(path, int(stat.Uid)+idpair.UID, int(stat.Gid)+idpair.GID)
```

Comment:
Might be clearer to use `i.idMapping.ToContainer`

Reply:
Ok switched to using `ToHost`, `ToContainer` is the other way around and was giving me errors. With `ToHost` the uid/gid is the right one inside the container. Might be a good idea to add a test for this

Reply:
Added a test for this